pkg/cli/sso.go

```diff
@@ -0,0 +1,216 @@
+		if err != nil {
+			return err
+		}
+
+		c.Writef("\n")
+	}
+
+	issuer := coalesce(c.String("issuer"), os.Getenv("SSO_ISSUER"))
+	if issuer == "" {
+		c.Writef("SSO ISSUER: ")
+
+		issuer, err = c.ReadSecret()
+		if err != nil {
+			return err
+		}
+
+		c.Writef("\n")
+	}
+
+	if err := c.SettingWrite("provider", provider); err != nil {
+		return err
+	}
+
+	if err := c.SettingWrite("client_id", clientID); err != nil {
+		return err
+	}
+
+	if err := c.SettingWrite("client_secret", clientSecret); err != nil {
+		return err
+	}
+
+	if err := c.SettingWrite("issuer", issuer); err != nil {
+		return err
+	}
+
+	return c.OK()
+}
+
+func SsoLogin(rack sdk.Interface, c *stdcli.Context) error {
+	provider, err := c.SettingRead("provider")
+	if err != nil {
+		return err
+	}
+
+	clientID, err := c.SettingRead("client_id")
+	if err != nil {
+		return err
+	}
+
+	clientSecret, err := c.SettingRead("client_secret")
+	if err != nil {
+		return err
+	}
+
+	issuer, err := c.SettingRead("issuer")
+	if err != nil {
+		return err
+	}
+
+	nonce, _ := sso.GenerateNonce()
+
+	p, err := sso.Initialize(provider, structs.SsoProviderOptions{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		Issuer:       issuer,
+		Scope:        "openid profile email",
+		State:        sso.GenerateState(),
+		Nonce:        nonce,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	server := &http.Server{Addr: "/authorization-code/callback"}
+
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		authCodeCallbackHandler(w, r, server, c, p)
+	})
+
+	port := fmt.Sprintf(":%s", "8090")
+	l, err := net.Listen("tcp", port)
+	if err != nil {
+		fmt.Printf("snap: can't listen to port %s: %s\n", port, err)
+		os.Exit(1)
+	}
+
+	log.Println("Waiting for login... ")
+
+	time.Sleep(1 * time.Second)
+	sso.Openbrowser(p.RedirectPath())
+	time.Sleep(1 * time.Second)
+
+	server.Serve(l)
+
+	return c.OK()
+}
+
+func authCodeCallbackHandler(w http.ResponseWriter, r *http.Request, server *http.Server, c *stdcli.Context, p structs.SsoProvider) {
+	if r.URL.Query().Get("state") != p.Opts().State {
+		fmt.Fprintln(w, "The state was not as expected")
+		return
+	}
+
+	if r.URL.Query().Get("code") == "" {
+		fmt.Fprintln(w, "The code was not returned or is not accessible")
+		return
+	}
+
+	exchange := p.ExchangeCode(r, r.URL.Query().Get("code"))
+	if exchange.Error != "" {
+		fmt.Fprintf(w, exchange.Error)
+		fmt.Fprintf(w, exchange.ErrorDescription)
+		return
+	}
+
+	verificationError := p.VerifyToken(exchange.IdToken)
+
+	if verificationError != nil {
+		fmt.Println(verificationError)
+	}
+
+	if err := c.SettingWrite("bearer_token", exchange.AccessToken); err != nil {
+		fmt.Fprintf(w, "Could not set Bearer Token on the CLI settings")
+		return
+	}
+
+	w.Header().Set("Content-Type", "text/html")
+	w.Write(openSuccessHTMLFile())
+
+	cleanup(server)
+}
+
+func openSuccessHTMLFile() []byte {
+	file, err := os.Open("../../sso/templates/success.html")
```

Comment:
this file won't be available in the user machine when they install the CLI, by default go doesn't include static files into the binary executable
to use the file you have to "pack" it inside the binary, we already have https://github.com/gobuffalo/packr in the project, you can use it to bundle the file and to be available in the CLI